hazelcast/src/hazelcast/client/compact.cpp

```diff
@@ -0,0 +1,678 @@
+        field_kind kind = descriptor.field_kind;
+        if (field_operations::get(kind).kind_size_in_byte_func() ==
+            field_kind_based_operations::VARIABLE_SIZE) {
+            variable_size_fields.push_back(&descriptor);
+        } else if (kind == field_kind::BOOLEAN) {
+            boolean_fields.push_back(&descriptor);
+        } else {
+            fixed_size_fields.push_back(&descriptor);
+        }
+    }
+
+    std::sort(
+      fixed_size_fields.begin(), fixed_size_fields.end(), kind_size_comparator);
+
+    int offset = 0;
+    for (auto descriptor : fixed_size_fields) {
+        descriptor->offset = offset;
+        offset += field_operations::get(descriptor->field_kind)
+                    .kind_size_in_byte_func();
+    }
+
+    int bit_offset = 0;
+    for (auto descriptor : boolean_fields) {
+        descriptor->offset = offset;
+        descriptor->bit_offset = bit_offset % util::Bits::BITS_IN_BYTE;
+        bit_offset++;
+        if (bit_offset % util::Bits::BITS_IN_BYTE == 0) {
+            offset += 1;
+        }
+    }
+    if (bit_offset % util::Bits::BITS_IN_BYTE != 0) {
+        offset += 1;
+    }
+
+    fixed_size_fields_length_ = offset;
+
+    int index = 0;
+    for (auto descriptor : variable_size_fields) {
+        descriptor->index = index;
+        index++;
+    }
+
+    for (auto& item : sorted_fields) {
+        field_definition_map_[item.first] = item.second;
+    }
+
+    number_of_var_size_fields_ = index;
+    schema_id_ = rabin_finger_print::fingerprint64(type_name_, sorted_fields);
+}
+
+int64_t
+schema::schema_id() const
+{
+    return schema_id_;
+}
+
+size_t
+schema::number_of_var_size_fields() const
+{
+    return number_of_var_size_fields_;
+}
+
+size_t
+schema::fixed_size_fields_length() const
+{
+    return fixed_size_fields_length_;
+}
+
+const std::string&
+schema::type_name() const
+{
+    return type_name_;
+}
+
+const std::unordered_map<std::string, field_descriptor>&
+schema::fields() const
+{
+    return field_definition_map_;
+}
+
+std::ostream&
+operator<<(std::ostream& os, const schema& schema)
+{
+    os << "type name " << schema.type_name() << ",number of var size fields "
+       << schema.number_of_var_size_fields() << ",fixed size fields length "
+       << schema.fixed_size_fields_length() << ",fields {";
+    for (const auto& item : schema.fields()) {
+        os << item.first << " " << item.second << ",";
+    }
+    os << "}";
+    return os;
+}
+
+std::ostream&
+operator<<(std::ostream& os, const field_descriptor& field_descriptor)
+{
+    os << field_descriptor.field_kind;
+    return os;
+}
+
+} // namespace pimpl
+
+namespace pimpl {
+schema_writer::schema_writer(std::string type_name)
+  : type_name(std::move(type_name))
+{}
+void
+schema_writer::add_field(std::string field_name, enum field_kind kind)
+{
+    field_definition_map.emplace(std::move(field_name),
+                                 field_descriptor{ kind, -1, -1, -1 });
```

Comment:
In Java, adding a field with the same field_name replaces it because `put` is used. Here if there is an existing mapping nothing will happen as I understand from the [docs](https://www.cplusplus.com/reference/unordered_map/unordered_map/emplace/). May this difference be for some users?

Reply:
Writing the same field twice should be illegal. I did not check it both in java and cpp because I thought no-one will do it, but may be we can discuss and a check later to both.

Reply:
We have decided to make behaviour same for cpp for now. Checks will be added later to java and others.